### PR TITLE
solve the unstability issues with TritonPart

### PR DESCRIPTION
**Pull request by ZhiangWang033**

### CODE REVIEW

**Comment by maliberty** [comment]
Needs to be clang-format'ed

File: src/par/src/TPPartitioner.cpp [hunk 1]
```diff
@@ -380,34 +367,109 @@ void TPpartitioner::OptimalPartCplexWarmStart(
...
                              0.0);
     unvisited_hyperedges.insert(std::pair<int, float>(e, score));
+    tot_cost += score;
   }
-
-  int max_num_hyperedges = 0;
-  std::vector<int> edge_mask;
-  float base_score = (*unvisited_hyperedges.begin()).second / 10.0;
+  // set the reserved total score for reduced hypergraph (100% for now)
+  const float reserved_tot_cost = tot_cost * 1.0;
```
Comment (maliberty) [inline src/par/src/TPPartitioner.cpp:373]: No need to * 1.0 but won't this lead to all edges being included and none pruned?

File: src/par/src/TPPartitioner.cpp [hunk 2]
```diff
@@ -380,34 +367,109 @@ void TPpartitioner::OptimalPartCplexWarmStart(
...
+    }
+    if (unique_partitions.size() == 1) {
+      const int part_id = *unique_partitions.begin();
+      for (int i = 0; i < num_parts_; i++) {
+        y[i][e] = (i == part_id) ? 1 : 0;
+      }            
+    } else {
+      for (int i = 0; i < num_parts_; i++) {
+        y[i][e] = 0;
+      }
```
Comment (maliberty) [inline src/par/src/TPPartitioner.cpp:451]: y was already initialized to 0 so this seems to be a no-op

File: src/par/src/TPPartitioner.cpp [hunk 3]
```diff
@@ -380,34 +367,109 @@ void TPpartitioner::OptimalPartCplexWarmStart(
...
+  for (auto e = 0; e < num_hyperedges; e++) {
+    std::set<int> unique_partitions;
+    for (auto& v : new_hyperedges[e]) {
+      unique_partitions.insert(solution[new_vertex_vec[v]]);
+    }
+    if (unique_partitions.size() == 1) {
+      const int part_id = *unique_partitions.begin();
+      for (int i = 0; i < num_parts_; i++) {
+        y[i][e] = (i == part_id) ? 1 : 0;
+      }
```
Comment (maliberty) [inline src/par/src/TPPartitioner.cpp:447]: would it suffice to do y[part_id][e] = 1; ?

File: src/par/src/TPPartitioner.h [hunk 4]
```diff
@@ -152,9 +152,9 @@ class TPpartitioner
   void RandomPart(const HGraph graph,
                   const matrix<float>& max_block_balance,
                   std::vector<int>& solution);
-  void OptimalPartCplex(const HGraph graph,
-                        const matrix<float>& max_block_balance,
-                        std::vector<int>& solution);
+  //void OptimalPartCplex(const HGraph graph,
+  //                      const matrix<float>& max_block_balance,
+  //                      std::vector<int>& solution);
```
Comment (maliberty) [inline src/par/src/TPPartitioner.h:157]: just remove

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/par/src/TPPartitioner.cpp (hunk 2 above)
Comment (ZhiangWang033) [reply src/par/src/TPPartitioner.cpp:451]: This line was written to make the algorithm more clear.  (If the hyperedge spans multiple partitions,  we need to set y to 0)

File: src/par/src/TPPartitioner.cpp (hunk 1 above)
Comment (ZhiangWang033) [reply src/par/src/TPPartitioner.cpp:373]: Leave 1.0 there.  We want to tune this parameter to better runtime and reasonable runtime later.

File: src/par/src/TPPartitioner.cpp (hunk 3 above)
Comment (ZhiangWang033) [reply src/par/src/TPPartitioner.cpp:447]: This was written to make the algorithm more clear.

File: src/par/src/TPPartitioner.h (hunk 4 above)
Comment (ZhiangWang033) [reply src/par/src/TPPartitioner.h:157]: I am okay with this.

**Comment by maliberty** [comment]
At a minimum you need to format the code
image

**Comment by ZhiangWang033** [comment]
Done. (I always forget to do this ....)


Thanks,
Zhiang

**Review by maliberty — APPROVED** [review]

**Comment by maliberty** [comment]
Looks like you need to update the unit tests:
image

**Comment by maliberty** [comment]
Replaced by https://github.com/The-OpenROAD-Project-private/OpenROAD/pull/1610

**Pull request closed without merging**